### PR TITLE
KAFKA-12889: log clean relative index range check of group consider empty log segment to avoid too many empty log segment left

### PR DESCRIPTION
**Pull request by iamgd67**
more detail see jira [KAFKA-12889](https://issues.apache.org/jira/browse/KAFKA-12889)

to avoid log index 4 byte relative offset overflow, log cleaner group check log segments offset to make sure group offset range not exceed Int.MaxValue.

this offset check currentlly not cosider next is next log segment is empty, so there will left empty log files every about 2^31 messages.

the left empty logs will be reprocessed every clean cycle, which will rewrite it with same empty content, witch cause little no need io.

for __consumer_offsets topic, normally we can set cleanup.policy to compact,delete to get rid of this.

my cluster is 0.10.1.1, but after aylize trunk code, it should has same problem too.

 

some of my left empty logs,(run ls -l)

rw-r---- 1 u g 0 Dec 16 2017 00000000000000000000.index
rw-r---- 1 u g 0 Dec 16 2017 00000000000000000000.log
rw-r---- 1 u g 0 Dec 16 2017 00000000000000000000.timeindex
rw-r---- 1 u g 0 Jan  15 2018 00000000002148249632.index
rw-r---- 1 u g 0 Jan  15 2018 00000000002148249632.log
rw-r---- 1 u g 0 Jan  15 2018 00000000002148249632.timeindex
rw-r---- 1 u g 0 Jan  27 2018 00000000004295766494.index
rw-r---- 1 u g 0 Jan  27 2018 00000000004295766494.log
rw-r---- 1 u g 0 Jan  27 2018 00000000004295766494.timeindex

### CODE REVIEW

**Comment by iamgd67** [comment]
@junrao could you please review this, failed tests seems unrelated.

**Comment by iamgd67** [comment]
@guozhangwang  could you please review this, thanks in advance.

**Comment by iamgd67** [comment]
@ijuma could you please review this pr, it has been several days since created, don't know what to continue.

File: core/src/test/scala/unit/kafka/log/LogCleanerTest.scala
Comment (showuon) [inline core/src/test/scala/unit/kafka/log/LogCleanerTest.scala]: Please add a space after comma

File: core/src/test/scala/unit/kafka/log/LogCleanerTest.scala
Comment (showuon) [inline core/src/test/scala/unit/kafka/log/LogCleanerTest.scala]: Could you please rephrase this comment? Not sure what are you trying to say.

File: core/src/test/scala/unit/kafka/log/LogCleanerTest.scala
Comment (showuon) [inline core/src/test/scala/unit/kafka/log/LogCleanerTest.scala]: space after comma. Thanks.

File: core/src/main/scala/kafka/log/LogCleaner.scala
Comment (showuon) [inline core/src/main/scala/kafka/log/LogCleaner.scala]: nit: add space between '=' `segs.head.size == 0`

File: core/src/main/scala/kafka/log/LogCleaner.scala
Comment (showuon) [inline core/src/main/scala/kafka/log/LogCleaner.scala]: nit: How about we say: `if first segment size is 0, we don't need to do the index offset range check`.

File: core/src/test/scala/unit/kafka/log/LogCleanerTest.scala
Comment (showuon) [inline core/src/test/scala/unit/kafka/log/LogCleanerTest.scala]: space after comma. Thanks.

**Review by showuon — COMMENTED** [review]
@iamgd67 , thanks for the patch. Left some comments. Thanks.

File: core/src/test/scala/unit/kafka/log/LogCleanerTest.scala
Comment (iamgd67) [reply core/src/test/scala/unit/kafka/log/LogCleanerTest.scala]: what I mean is

> because index file uses 4 byte relative index offset and current segments all none empty, segments will not group even their size is very small.

updated the comments

File: core/src/main/scala/kafka/log/LogCleaner.scala
Comment (iamgd67) [reply core/src/main/scala/kafka/log/LogCleaner.scala]: changed as your suggestion

**Comment by iamgd67** [comment]
@showuon thanks for your detailed review, changed and pushed.

**Review by showuon — APPROVED** [review]
LGTM! Thank you!

**Comment by iamgd67** [comment]
@showuon failed tests seem unrelated.

**Comment by showuon** [comment]
@iamgd67 , yes, the failed tests are flaky.
Next step is to wait for committers to review this PR **patiently**. Committers are willing to review, but they are also busy on their own daily jobs. Please stop emailing to whole dev group for PR review. Ping some committers that are familiar to this area, or some guys that did changes to this file before. Then, wait. Thank you. :)

```
    Build / JDK 8 and Scala 2.12 / kafka.server.RaftClusterTest.testCreateClusterAndCreateAndManyTopicsWithManyPartitions()
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.common.network.SslTransportLayerTest.[1] tlsProtocol=TLSv1.2, useInlinePem=false
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.common.network.SslTransportLayerTest.[2] tlsProtocol=TLSv1.2, useInlinePem=true
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.common.network.SslTransportLayerTest.[3] tlsProtocol=TLSv1.3, useInlinePem=false
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.common.network.SslTransportLayerTest.[1] tlsProtocol=TLSv1.2, useInlinePem=false
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.common.network.SslTransportLayerTest.[2] tlsProtocol=TLSv1.2, useInlinePem=true
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.common.network.SslTransportLayerTest.[3] tlsProtocol=TLSv1.3, useInlinePem=false
    Build / JDK 11 and Scala 2.13 / kafka.server.RaftClusterTest.testCreateClusterAndCreateAndManyTopics()
    Build / JDK 11 and Scala 2.13 / kafka.server.RaftClusterTest.testCreateClusterAndCreateAndManyTopics()
```

**Comment by guozhangwang** [comment]
@iamgd67 Thanks for the PR! and also thank @showuon for the reviews. I made a quick pass and it LGTM. Merging to trunk now.

**Pull request merged**
